src/gui/shell.cpp

```diff
@@ -83,10 +83,12 @@ void Shell::fontError(const QString& msg)
 QString Shell::fontDesc()
 {
 	QString fdesc = QString("%1:h%2").arg(fontFamily()).arg(fontSize());
-	if (font().bold()) {
+	if (font().weight() == QFont::Bold) {
 		fdesc += ":b";
 	} else if (font().weight() == QFont::Light) {
 		fdesc += ":l";
+	} else if (font().weight() == QFont::DemiBold) {
+		fdesc += ":sb";
```

Comment:
I think we should end with a check to `font().bold()` or `font().weight() > QFont::Medium`, and add a comment.

```C++
// Must be checked last: fallback to bold if weight is greater than Medium.
else if (font().bold()) { 
    fdesc += :b;
}
```

This is safer because `::weight()` can return values in between the enumerations. Function `::bold()` actually performs the check `::weight() > QFont::Medium`. That would more gracefully handle in-between values.

Sorry for the nitpick... I haven't observed any issues but it would be safer code.

Reply:
I thought of that, but I think that will cause unintended weight issues.  If the font they chose is not weight=bold, then detecting is a bold is going to produce a different result.  After all, you don't do the reverse of this for light either.

Reply:
Sort of...

Let's say the user creates a font of weight 70.

That would fail each check currently in place:
- QFont::Bold == 75
- QFont::DemiBold == 63
- QFont::Light == 25

The existing master behavior would be to mark a font of weight 70 as `:b`, bold. This PR would leave the font unmarked.

With `::bold()` at the end, it checks `70 > QFont::Medium` or `70 > 57`. So the font is at least marked as bold. Not entirely correct, but better than doing nothing and not a regression.

Alternatively, you could properly handle the font ranges. We should do something for in-between values I think

Reply:
I undertand that.  And then when they select the font with ":b" they wont get the font they created!  How confusing is that?   It's not like you check if weight() < Medium and default to Light.

Reply:
Sorry for being pedantic, from my testing everything works great...

However, I could see this causing issues for users with obscure configurations in the future :)

Reply:
I don't follow... Can you elaborate with a specific example?

Here is how I understand the behavior:

**Before:**
weight == Light (25) -> Light
weight > Medium (57) -> Bold
weight <= Medium && weight != Light (25) -> Normal

99 -> Bold (not perfect, but reasonable)

**After:**
weight == Light (25) -> Light
weight == DemiBold (63) -> SemiBold
weight == Bold (75) -> Bold
[None of the above] -> Normal

99 -> Normal (behavior change, regression worse than before)

IMO, the "perfect" ranges are:
```C++
// Light
if (weight <= QFont::Light) {
	fdesc += ":l";
}
// Normal
else if (weight < DemiBold) {
}
// DemiBold
else if (weight < Bold) {
	fdesc += ":sb";
}
// Bold
else {
	fdesc += ":b";
}
```

However, that is rather confusing code. For readability's sake, I you could append `else if (font().bold())` as a reasonable compromise. Yes, that would result in odd behavior for weight range (Medium, DemiBold].

Sorry for being pedantic... Please let me know if I am missing something :)

Reply:
I understand what you're saying, I just think it should be a bijection.

Reply:
I suppose we simply disagree :man_shrugging:

Opinion: The behavior is less optimal than what was there before. For example, neovim-qt will now display a regular font for heavy weights such as `Black`. Font weight is graded, so it would not surprise me if this causes issues with `:b` on some systems...

Reply:
If your ranges interpretation was the right one, then why wouldn't they have coded ranges into Qt?

Reply:
Font weight is ranged. Why else would QFont::bold() be implemented this way?
``` C++
inline bool QFont::bold() const
{ return weight() > Medium; }
```

Qt just doesn't provide the exact API we need in this application. This is a very specific use case:

1. We are doing odd things: converting QFont to description and back.
2. We don't support every style enumeration.
3. We aren't actually storing the weight directly.

I'm open to solutions... But I think we should handle in-between values. Previously, the behavior made sense given the supported enumerations

Maybe we can add a `:w{num}` for in-between weights? I remember seeing a bug actually requesting this...

Other Issues: #532 #597

Reply:
Maybe we're spending way too long on something no one is ever going to see?  Would you mind just sending me the code you want me to use, or else checking in the version you like yourself?

Reply:
Sorry, I'm not trying to be a PITA here :/

You are making a change to the bold logic that is not strictly required. If you want to do that great!

However, you need to prove that your new code will work for all users. I haven't seen this yet...

> Would you mind just sending me the code you want me to use

I've made a few suggestions, see my first comment.

Here is the code in-place:
```C++
QString Shell::fontDesc()
{
	QString fdesc = QString("%1:h%2").arg(fontFamily()).arg(fontSize());
	if (font().weight() == QFont::Bold) {
		fdesc += ":b";
	} else if (font().weight() == QFont::Light) {
		fdesc += ":l";
	} else if (font().weight() == QFont::DemiBold) {
		fdesc += ":sb";
	}
	// Must be checked last: fallback to bold if weight is greater than Medium.
	else if (font().bold()) { 
    		fdesc += :b;
	}
```

If you don't like that, I understand. There is also the range based solution above, or the option to add arbitrary weight `:w` (requested by other users, best solution more work?).

Thank you for your contributions, this is a valuable feature requested by multiple users!

Reply:
Don't worry, you're not being a pain.  The w solution sounds best.  Will send you a merged changelist this weekend.

Reply:
Just to confirm, I think the logical thing to do is get rid of "sb" then and keep it simple: b, l, wX for X a number between 0 and 100

Reply:
Agreed, I like the `:w` solution best too. Thanks!

I don't see any harm in keeping `:sb` around, it seems like a useful option. If you want to remove it, that's fine too.